### PR TITLE
Show 'Type' for PVCs in the Mounted Volumes list

### PR DESCRIPTION
**Pull request by AyushAmbastha**
Before-

BeforePod

After-

AfterPod

Note: The description I've given for PVCs in the pods.tsx file is -
'A Persistent Volume Claim is a request for a resource i.e A Persistent Volume'

### CODE REVIEW

File: frontend/public/module/k8s/pods.ts [hunk 1]
```diff
@@ -82,6 +82,11 @@ export const VolumeSource = {
     label: 'Projected',
     description: 'A projected volume maps several existing volume sources into the same directory.',
   },
+  persistentVolumeClaim: {
+    id: 'persistentVolumeClaim',
+    label: 'Persistent Volume Claim',
+    description: 'A Persistent Volume Claim is a request for a resource i.e A Persistent Volume',
```
Comment (spadgett) [inline frontend/public/module/k8s/pods.ts]: Maybe just

```suggestion
    description: 'A Persistent Volume Claim is a request for a Persistent Volume',
```

File: frontend/public/components/utils/volume-icon.jsx [hunk 2]
```diff
@@ -3,15 +3,23 @@ import * as React from 'react';
...
+    [VolumeSource.persistentVolumeClaim.id, 'co-m-resource-persistentvolumeclaim'],
+  ]);
   const faClass = faClasses[kind];
+  const coClass = coClasses[kind];
 
-  return <span className="co-icon-and-text co-m-volume-icon">
-    {faClass && <i className={`fa ${faClass} co-icon-and-text__icon`}></i>}
+  return <span className="co-icon-and-text">
+    {faClass && <i className={`fa co-m-volume-icon ${faClass} co-icon-and-text__icon`}></i>}
+    {coClass && <span className={`co-m-resource-icon ${coClass}`} title={kind}>{kindMapping[kind]}</span>}
```
Comment (spadgett) [inline frontend/public/components/utils/volume-icon.jsx]: I don't think we should use the PVC resource icon here since we don't use the `S` for secret volumes or the `CM` for config map volumns.

Comment (AyushAmbastha) [reply frontend/public/components/utils/volume-icon.jsx]: @julim@redhat.com

File: frontend/public/module/k8s/pods.ts (hunk 1 above)
Comment (AyushAmbastha) [reply frontend/public/module/k8s/pods.ts]: Okay! :) Ack.

File: frontend/public/components/utils/volume-icon.jsx (hunk 2 above)
Comment (spadgett) [reply frontend/public/components/utils/volume-icon.jsx]: Typically we use it for a specific resource instance with a link as well, so we're breaking the pattern.

Comment (AyushAmbastha) [reply frontend/public/components/utils/volume-icon.jsx]: Which icon do you suggest we use then?

Comment (spadgett) [reply frontend/public/components/utils/volume-icon.jsx]: Honestly, I'm not sure having a different icon for every kind of volume (emptydir, hostPath, secret, config map, PVC, etc.) scales. It would be difficult to find meaningful icons for all of these. We might just leave it off.

If it links to a specific secret, config map, or PVC, it would be nice to have a link to that specific object with the resource icon badge.

**Comment by spadgett** [comment]
Hi, I wanted to check on the status of this PR. I think the simple thing to do is to add a `ResourceLink` for all types that support it (PVC, secret, config map), and use a plain text value for other types.

**Comment by AyushAmbastha** [comment]
> Hi, I wanted to check on the status of this PR. I think the simple thing to do is to add a `ResourceLink` for all types that support it (PVC, secret, config map), and use a plain text value for other types.

Hey @spadgett , Do you still feel that the Icon such as the 'PVC' one is not required and that Adding `ResourceLink` for all the ones that are supported, is enough?

**Comment by AyushAmbastha** [comment]
Screenshot from 2019-05-27 15-25-44

@spadgett This is how it looks like after using `ResourceLink`. 
Also, the 'Name' column deals with the group names for volume types such as PVCs etc , but not for Secrets. So the secret name will be displayed twice. Then again, the link cant be put on the 'Name' column as it wouldn't be correct in the case of Volumes and PVCs. 
What do you think we should do in this case?

**Comment by spadgett** [comment]
Volume name isn't always the same as the secret name, so we should display both.

The links should be like

```
(PVC) my-pvc
(S) my-secret
```

Make the actual name of the object the link text. This is consistent with other links throughout the console.

Thanks!

**Comment by AyushAmbastha** [comment]
Screenshot from 2019-05-28 14-01-09

Screenshot from 2019-05-28 14-01-51

@spadgett Made the changes as requested, Please review. For volume types in which neither the Resource link exists nor the `className=fa-...` is chosen, nothing is displayed and null is returned. Is that okay or `-` should be displayed?

Referring to line - https://github.com/openshift/console/pull/1344/files#diff-9f09cd1fb4d8499069d32d298cb76de7R21

File: frontend/public/components/utils/volume-icon.jsx [hunk 3]
```diff
@@ -1,17 +1,22 @@
 import * as _ from 'lodash-es';
 import * as React from 'react';
 import {VolumeSource} from '../../module/k8s/pods';
+import { ResourceLink } from './resource-link';
 
-export const VolumeIcon = ({kind}) => {
+export const VolumeIcon = ({kind, name}) => {
```
Comment (spadgett) [inline frontend/public/components/utils/volume-icon.jsx]: We should rename this. This is no longer just an icon. It should be a component that renders the entire value, not just the icon now (since we have to render links in some cases).

File: frontend/public/module/k8s/pods.ts [hunk 4]
```diff
@@ -51,6 +51,7 @@ export const VolumeSource = {
   },
   secret: {
     id: 'secret',
+    link: 'Secret',
```
Comment (spadgett) [inline frontend/public/module/k8s/pods.ts]: s/link/kind

File: frontend/public/module/k8s/pods.ts [hunk 5]
```diff
@@ -101,12 +109,18 @@ const genericFormatter = volInfo => {
...
+    if (key === 'optional') {
+      return '';
+    }
     return volInfo[key];
   });
   if (keys.indexOf('readOnly') !== -1) {
     parts.push(volInfo.readOnly ? 'ro' : 'rw');
   }
-  return parts.join(' ') || null;
+  return parts.join(' ').trim() || null;
```
Comment (spadgett) [inline frontend/public/module/k8s/pods.ts]: Why is `trim` necessary?

Comment (AyushAmbastha) [reply frontend/public/module/k8s/pods.ts]: For resource types such as Secrets and ConfigMaps, there is another parameter returned called `defaultMode` which is a value denoting its permissions. In the case of a secret that I had in my cluster, the parameter was `defaultMode : 420`
So the `loc` returned was `420 <secret_name>` , On omitting the defaultMode, the value returned was ` <secret_name>` with a space in the beginning. This led to the path being incorrect and hence, the link wasn't working.

File: frontend/public/components/utils/volume-icon.jsx (hunk 3 above)
Comment (AyushAmbastha) [reply frontend/public/components/utils/volume-icon.jsx]: Yes, I agree. I didn't want to change the name without asking you first. What do you suggest we should rename it to?

File: frontend/public/components/utils/volume-icon.jsx [hunk 6]
```diff
@@ -1,17 +1,22 @@
...
+  const volumeLink = _.get(VolumeSource[kind], 'link', false);
 
-  return <span className="co-icon-and-text co-m-volume-icon">
+  if (faClass) return <span className="co-icon-and-text co-m-volume-icon">
     {faClass && <i className={`fa ${faClass} co-icon-and-text__icon`}></i>}
     {_.get(VolumeSource[kind], 'label', '')}
-  </span>;
+  </span>
+
+  if (name) return <ResourceLink kind={volumeLink} name={name}/>
```
[the remark this answers is not in the document]
Comment (AyushAmbastha) [reply frontend/public/components/utils/volume-icon.jsx]: Ack!

File: frontend/public/components/utils/volume-icon.jsx [hunk 7]
```diff
@@ -1,17 +1,26 @@
...
 
-export const VolumeIcon = ({kind}) => {
+export const VolumeLink = ({kind, name, namespace}) => {
   const faClasses = _.fromPairs([
     [VolumeSource.emptyDir.id, 'fa-folder-open-o'],
     [VolumeSource.hostPath.id, 'fa-files-o'],
-    [VolumeSource.secret.id, 'fa-lock'],
   ]);
   const faClass = faClasses[kind];
+  const link = _.get(VolumeSource[kind], 'link', false);
```
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/utils/volume-icon.jsx]: We're calling this `link` but really it's a k8s kind? We should give this a better name. Maybe `k8sKind` and `volumeSourceKind` for the other variable to avoid confusion.

File: frontend/public/components/utils/volume-icon.jsx [hunk 8]
```diff
@@ -1,17 +1,26 @@
 import * as _ from 'lodash-es';
 import * as React from 'react';
 import {VolumeSource} from '../../module/k8s/pods';
+import { ResourceLink } from './resource-link';
 
-export const VolumeIcon = ({kind}) => {
+export const VolumeLink = ({kind, name, namespace}) => {
```
Comment (spadgett) [inline frontend/public/components/utils/volume-icon.jsx]: Maybe call this `VolumeType` to match the table heading.

We should also rename this file if we're renaming the component.

File: frontend/public/components/utils/volume-icon.jsx [hunk 9]
```diff
@@ -1,17 +1,26 @@
 import * as _ from 'lodash-es';
 import * as React from 'react';
 import {VolumeSource} from '../../module/k8s/pods';
+import { ResourceLink } from './resource-link';
```
Comment (spadgett) [inline frontend/public/components/utils/volume-icon.jsx]: Use consistent spaces inside braces for imports.

File: frontend/public/module/k8s/index.ts [hunk 10]
```diff
@@ -242,8 +242,17 @@ export type PodStatus = {
   [key: string]: any;
 };
 
+export type PodMetadata ={
+  name: string,
```
Comment (spadgett) [inline frontend/public/module/k8s/index.ts]: These lines should end with `;` and not `,`

File: frontend/public/module/k8s/index.ts [hunk 11]
```diff
@@ -242,8 +242,17 @@ export type PodStatus = {
   [key: string]: any;
 };
 
+export type PodMetadata ={
```
Comment (spadgett) [inline frontend/public/module/k8s/index.ts]: We already have a common metadata definition. No need to redefine it.

https://github.com/openshift/console/blob/master/frontend/public/module/k8s/index.ts#L33-L41

**Comment by AyushAmbastha** [comment]
@spadgett All changes requested have been taken care of. Please review.

File: frontend/public/components/volumes-table.tsx [hunk 12]
```diff
@@ -56,6 +56,7 @@ const VolumeRow: React.FC<VolumeRowProps> = ({value, pod}) => {
   const kind = _.get(getVolumeType(value.volume), 'id', '');
   const loc = getVolumeLocation(value.volume);
   const name = value.name;
+  const namespace = pod.metadata.namespace;
```
Comment (spadgett) [inline frontend/public/components/volumes-table.tsx]: You want to make sure you're getting the namespace from the original resource, not the pod template. I don't think it's always set here.

File: frontend/public/module/k8s/pods.ts [hunk 13]
```diff
@@ -84,6 +86,12 @@ export const VolumeSource = {
     label: 'Projected',
     description: 'A projected volume maps several existing volume sources into the same directory.',
   },
+  persistentVolumeClaim: {
```
Comment (spadgett) [inline frontend/public/module/k8s/pods.ts]: It's unclear to me why this change is needed. I'm not sure it's right since there are other more specific values for different kinds of PVCs above.

File: frontend/public/module/k8s/pods.ts (hunk 4 above)
Comment (spadgett) [reply frontend/public/module/k8s/pods.ts]: This comment is outstanding.

**Comment by anmolsachan** [comment]
@spadgett The contributor to this PR is no longer contributing to the project. Is this pr required or all the features are beung handled in #1891 ?

**Comment by anmolsachan** [comment]
@zherman0 The contributor to this PR is no longer contributing to the project. Is this pr required or all the features are being handled in #1891 ?

**Comment by zherman0** [comment]
@anmolsachan - [PR1891](https://github.com/openshift/console/pull/1891) adds a kebab menu to the volume table and allows for the `Remove Options`.  In doing so, the volume tables were refactored to use the new PF style.  
That said, no additional work was done on types in the table and no work was done with any volume icons.
I think this PR is still needed if this work is considered needed, but it will need to be reworked a little once 1891 merges.

**Comment by spadgett** [comment]
Closing in favor of #2107

/close
